File: python/size.py
from __future__ import annotations
from position import Position


class Size:
    def __init__(self, num_rows: int, num_cols: int):
        self._num_rows = num_rows
        self._num_cols = num_cols

    def copy(self):
        return Size(self._num_rows, self._num_cols)

    def is_within_bounds(self, position: Position) -> bool:
        return (position.row >= 0
                and position.row < self._num_rows
                and position.col >= 0
                and position.col < self._num_cols)

    def __str__(self):
        return f"{self._num_rows} rows by {self._num_cols} cols"

    def __eq__(self, other: Size) -> bool:
        return self.num_cols == other.num_cols and self.num_rows == other.num_rows

    @property
    def num_rows(self) -> int:
        return self._num_rows

    @property
    def num_cols(self) -> int:
        return self._num_cols

File: rust/src/size.rs
use crate::{enums::Direction, position::Position};

#[derive(Copy, Clone)]
pub struct Size {
    pub num_rows: usize,
    pub num_cols: usize,
}

impl Size {
    pub fn new(num_rows: usize, num_cols: usize) -> Size {
        Size { num_rows, num_cols }
    }

    pub fn is_within_bounds(&self, position: &Position) -> bool {
        position.row >= 0
            && position.row < self.num_rows
            && position.col >= 0
            && position.col < self.num_cols
    }

    pub fn increment(&self, position: &Position, direction: Direction) -> Option<Position> {
        match direction {
            Direction::Left => {
                if position.row == 0 {
                    return None;
                }
                return Some(Position::new(position.row - 1, position.col));
            }
            Direction::Right => {
                if position.row >= (self.num_rows - 1) {
                    return None;
                }
                return Some(Position::new(position.row + 1, position.col));
            }
            Direction::Up => {
                if position.col == 0 {
                    return None;
                }
                return Some(Position::new(position.row, position.col - 1));
            }
            Direction::Down => {
                if position.row >= (self.num_cols - 1) {
                    return None;
                }
                return Some(Position::new(position.row, position.col + 1));
            }
        }
    }
}
